include/boost/geometry/algorithms/detail/disjoint/linear_areal.hpp

```diff
@@ -139,66 +227,127 @@ struct disjoint_segment_areal
+                Segment, unary_predicate_type
             >::apply(boost::begin(interior_rings),
                      boost::end(interior_rings),
                      unary_predicate_type(segment, strategy));
+
+        return res;
     }
 
 
 public:
     template <typename IntersectionStrategy>
-    static inline bool apply(Segment const& segment,
-                             Polygon const& polygon,
-                             IntersectionStrategy const& strategy)
+    static inline intersection_return_type
+    apply(Segment const& segment,
+          Polygon const& polygon,
+          IntersectionStrategy const& strategy)
     {
         typedef typename geometry::ring_type<Polygon>::type ring;
 
-        if ( !disjoint_range_segment_or_box
-                 <
-                     ring, closure<Polygon>::value, Segment
-                 >::apply(geometry::exterior_ring(polygon), segment, strategy) )
+        intersection_return_type res = disjoint_range_segment_or_box_with_info
+        <
+            ring, closure<Polygon>::value, Segment
+        >::apply(geometry::exterior_ring(polygon), segment, strategy);
+
+        if ( res.count != 0 )
         {
-            return false;
+            return res;
         }
 
-        if ( !check_interior_rings(geometry::interior_rings(polygon), segment, strategy) )
+
+        res = check_interior_rings(geometry::interior_rings(polygon),
+                                   segment,
+                                   strategy);
+
+        if ( res.count != 0 )
         {
-            return false;
+            return res;
         }
 
         typename point_type<Segment>::type p;
         detail::assign_point_from_index<0>(segment, p);
 
-        return !geometry::covered_by(p, polygon,
-                    strategy.template get_point_in_geometry_strategy<Segment, Polygon>());
+        if (geometry::covered_by(p, polygon,
+                    strategy.template get_point_in_geometry_strategy<Segment, Polygon>()))
+        {
+            intersection_return_type res;
+            res.count = 1;
+            res.intersections[0] = p;
+            return res;
+        }
+
+        intersection_return_type res_empty;
+        res_empty.count = 0;
+        return res_empty;
+
     }
 };
 
 
+template <typename Segment, typename Polygon>
+class disjoint_segment_areal<Segment, Polygon, polygon_tag>
+{
+
+public:
+    template <typename IntersectionStrategy>
+    static inline bool apply(Segment const& segment,
+                             Polygon const& polygon,
+                             IntersectionStrategy const& strategy)
+    {
+        return disjoint_segment_areal_with_info<Segment, Polygon>
+                ::apply(segment, polygon, strategy).count == 0;
+    }
+};
+
+
+template <typename Segment, typename MultiPolygon>
+struct disjoint_segment_areal_with_info<Segment, MultiPolygon, multi_polygon_tag>
+{
+    typedef typename point_type<Segment>::type p_type;
```

Comment:
Here again, point type taken from Segment (first geometry) and I'm not sure if that's correct.

Reply:
changed to most precise